### PR TITLE
refactor: remove internal navigation controller

### PR DESCRIPTION
**Pull request by MarshallOfSound**
There's an unanswered question here with a potential change in "reload" behavior.  It's a toss up between "the old behavior was a bug" and "the new behavior is potentially unsafe / also a bug".  See the implementation of `WebContents::Reload`

Notes: no-notes

### CODE REVIEW

File: shell/browser/api/electron_api_web_contents.h [hunk 1]
```diff
@@ -154,17 +154,26 @@ class WebContents : public gin::Wrappable<WebContents>,
...
+  void Reload();
+  void ReloadIgnoringCache();
   void DownloadURL(const GURL& url);
   GURL GetURL() const;
   std::u16string GetTitle() const;
   bool IsLoading() const;
   bool IsLoadingMainFrame() const;
   bool IsWaitingForResponse() const;
   void Stop();
-  void ReloadIgnoringCache();
```
Comment (MarshallOfSound) [inline shell/browser/api/electron_api_web_contents.h:164]: So I found out while doing this that this header declaration has existed for 6 years and pointed at, well, nothing lol

**Comment by pushkin-** [comment]
Not to take over your refactor PR, but [here's](https://github.com/electron/electron/issues/28338) an issue that could maybe be fixed up that relates to the navigation controller :)

**Comment by MarshallOfSound** [comment]
@pushkin- That issue partially seems like expected behavior to me and is also unrelated to the navigation controller implementation.  Specifically the code that powers that is the implementation of `loadURL` which is Electron specific and independent of the chromium navigation controller.

**Review by deepak1556 — APPROVED** [review]

File: shell/browser/api/electron_api_web_contents.cc [hunk 2]
```diff
@@ -2007,18 +2023,64 @@ void WebContents::Stop() {
...
 }
 
+bool WebContents::CanGoBack() const {
+  return web_contents()->GetController().CanGoBack();
+}
+
 void WebContents::GoBack() {
+  if (!CanGoBack())
+    return;
   web_contents()->GetController().GoBack();
```
Comment (nornagon) [inline shell/browser/api/electron_api_web_contents.cc:2033]: nit: 
```suggestion
  if (CanGoBack())
    web_contents()->GetController().GoBack();
```

reads a bit easier. same for the other methods here.

File: shell/browser/api/electron_api_web_contents.cc [hunk 3]
```diff
@@ -2007,18 +2023,64 @@ void WebContents::Stop() {
...
 void WebContents::GoToOffset(int offset) {
+  if (!CanGoToOffset(offset))
+    return;
   web_contents()->GetController().GoToOffset(offset);
 }
 
+bool WebContents::CanGoToIndex(int index) const {
+  if (index < 0)
+    return false;
+  return index < GetHistoryLength();
```
Comment (nornagon) [inline shell/browser/api/electron_api_web_contents.cc:2059]:
```suggestion
  return index >= 0 && index < GetHistoryLength();
```

File: shell/browser/api/electron_api_web_contents.cc [hunk 4]
```diff
@@ -1973,6 +1976,23 @@ void WebContents::LoadURL(const GURL& url,
   NotifyUserActivation();
 }
 
+// TODO(MarshallOfSound): Figure out what we need to do with post data here, I
+// believe the default behavior when we pass "true" is to phone out to the
+// delegate and then the controller expects this method to be called again with
+// "false" if the user approves the reload.  For now this would result in
+// ".reload()" calls on POST data domains failing silently.  Passing false would
+// result in them succeeding, but reposting which although more correct could be
+// considering a breaking change.
```
Comment (nornagon) [inline shell/browser/api/electron_api_web_contents.cc:1985]: should we expose this as a param?

Comment (MarshallOfSound) [reply shell/browser/api/electron_api_web_contents.cc:1985]: I think we should, but that also (for correct implementations) would require us exposing the delegate event `did-get-post-data-reload-request` or something 🤔  I think for now defaulting to the safe value and exposing the new API later is OK.

Thoughts?

Comment (nornagon) [reply shell/browser/api/electron_api_web_contents.cc:1985]: sgtm

**Pull request merged**